Ability to pass arbitrary inputs from init_printing to the printer

jbm950 opened the pull request:
This code should allow arbitrary inputs to be passed from init_printing
to the printer of choice.

Example:

```
>>> from sympy import *
>>> init_printing(pretty_printing=False, use_latex=True, mode="inline")
>>> Rational(7, 2)
$\frac{7}{2}$
```

The init_printing in this case passes mode="inline" to the latex printer
because mode is not one of its keyword arguments and gets passed on
through **settings. This could possibly reduce the need for a lot of the
other keyword inputs for the init_printing function, however, it seemed
like the Ipython interactive printing code used those and that clean up
would take more investigation.

Note that when not using Ipython there is not currently code that
supports choosing the latex printer from init_printing. I plan to make a
pull request and open an issue immediately following this one, however,
that should fix this.

Note: I do not currently have Ipython with which to test these changes

This should close #6711

jbm950 committed:
Ability to pass arbitrary inputs from init_printing to the printer

This code should allow arbitrary inputs to be passed from init_printing
to the printer of choice.

Example:

>>> from sympy import *
>>> init_printing(pretty_printing=False, use_latex=True, mode="inline")
>>> Rational(7, 2)
$\frac{7}{2}$

The init_printing in this case passes mode="inline" to the latex printer
because mode is not one of its keyword arguments and gets passed on
through **settings. This could possibly reduce the need for a lot of the
other keyword inputs for the init_printing function, however, it seemed
like the ipython interactive printing code used those and that clean up
would take more investigation.

Note that when not using Ipython there is not currently code that
supports choosing the latex printer from init_printing. I plan to make a
pull request and open an issue immediately following this one, however,
that should fix this.

jbm950 commented:
This will probably also require a documentation update. Should the documentation update be solely in the docstring of init_printing?

asmeurer commented:
You should definitely install the notebook so you can test this. The easiest way is to install Anaconda.

smichr commented:
> I plan to make a
> pull request and open an issue immediately following this one, however,
> that should fix this.

How does the closing of #10835 affect this PR?

jbm950 commented:
The closing of #10835 only affects the example that I gave because latex is not a terminal printer that init_printing will use. Settings will still pass to the printers that init_printing uses in the terminal. Therefore this PR is still separate and is still being worked on. I currently have access to IPython and am testing the passing of settings to the IPython printing portion of code.

smichr commented:
OK, thanks for the update. I don't know a lot about these options so was only checking to make sure this is active. I will let someone else deal with the review.

jbm950 commented:
I'm currently running into a road block with looking into passing options to latex in the ipython printer function. _init_ipython_printing() sets three different formatters for the same list of printable_types: _print_plain() which uses pretty as its stringify_func, _print_latex_text() which calls the sympy latex printer directly and _print_latex_png(). When using the qtconsole and inputting something like Rational(6, 7) all three of the previously listed functions are called. I have tried passing options directly to the latex call in _print_latex_text() and can prove that it is being passed correctly because it will throw errors if I misspell one of the keyword inputs but not when they're entered correctly. Despite being able to pass these, there seems to be no effect on the output that it displays. I'm guessing this is on the IPython side on how it handles inputs as it passes around its formatters and renders the latex for output. 

Note: When IPython calls the three formatter functions it does not appear to do so in a consistent order.

Summary: 
*  I have the settings being passed into the _init_ipython_printing() but I am not sure where to pass them to have any effect once inside the function.
* Settings are being properly passed to the printer used in a regular terminal

asmeurer commented:
To affect latex you'll need to add the settings to the places in the code that call `latex()`.

jbm950 commented:
Ok there were about 3 additional calls to `latex()` that I had missed. The change currently works (I tested with `fold_frac_powers=True` and obtained the desired change in output. Writing up a commit now.

jbm950 commented:
The latex calls in the IPython printer have specific modes already set. To compensate for this I use settings.pop("mode", None) and want to display a message letting the user know that mode was not allowed to be specified. Should this message be an error, a simply message printed to the screen or should no message be given at all since the latex is going to be rendered and so the user won't notice that it didn't use their specified mode?

jbm950 committed:
Allow passing of arguments to latex printer in IPython

The ability to pass settings to the latex printer in the IPython code of
init_printing() has been added. Now latex can be given additional
arguments when working in IPython such as fold_frac_powers.

Example

In [1]: from sympy import *

In [2]: init_printing(fold_frac_powers=True)

Also added is a portion of code that removes any attempt to user specify
the mode of latex (example: mode="inline"). This is because the calls to
latex in the IPython printer already specify modes. The warning that
'mode' has been removed from the settings dictionary is sent using the
debug call just like the other portions of the init_printing code.

asmeurer commented:
I think a debug call is fine for mode.
